plugins/woocommerce/src/Internal/DependencyManagement/ServiceProviders/ProductQueryFiltersServiceProvider.php

```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Automattic\WooCommerce\Internal\DependencyManagement\ServiceProviders;
+
+use Automattic\WooCommerce\Internal\ProductQueryFilters\FilterDataProvider;
+use Automattic\WooCommerce\Internal\ProductQueryFilters\FilterClausesGenerator;
+use Automattic\WooCommerce\Internal\ProductQueryFilters\MainQueryFilters;
+
+/**
+ * LoggingServiceProvider class.
+ */
+class ProductQueryFiltersServiceProvider extends AbstractInterfaceServiceProvider {
+	/**
+	 * List services provided by this class.
+	 *
+	 * @var string[]
+	 */
+	protected $provides = array(
+		FilterClausesGenerator::class,
+		FilterDataProvider::class,
+		MainQueryFilters::class,
+	);
+
+	/**
+	 * Registers services provided by this class.
+	 *
+	 * @return void
+	 */
+	public function register() {
+		$this->share( FilterClausesGenerator::class );
+		$this->share_with_implements_tags( MainQueryFilters::class )->addArgument( FilterClausesGenerator::class );
+		$this->add( FilterDataProvider::class )->addArgument( FilterClausesGenerator::class );
```

Comment:
The reason I use `add` here instead of `share` is I want to allow changing the clauses generator later, I don't want to set the generator every time we use it. But it comes with the cost, `FilterClausesGenerator` is not cached. What do you think? @Konamiman @vedanshujain

Reply:
I'm not sure if I understand what you mean, by "changing the clauses generator later" you mean retriving a different instance from the container? In that case you should register `FilterClausesGenerator` with `add` (not `share`), and then register `FilterDataProvider` with `share` but then use `wc_get_container` to get the instance of `FilterClausesGenerator` every time you need it (so: `FilterDataProvider` will still be single instance, but `FilterClausesGenerator` will be a different instance every time). Or maybe you mean something different?

Reply:
@Konamiman for example, I want to refactor the StoreAPI/ProductQueryFilters to use the new services, StoreAPI has its own clauses generator, so I use `set_filter_clauses_provider` to set the generator at run time.

https://github.com/woocommerce/woocommerce/blob/9b1d17997bf33475a1f814491547794ed1420539/plugins/woocommerce/src/StoreApi/Utilities/ProductQueryFilters.php#L17-L20

(`FilterData` is the old name of `FilterDataProvider`, same for `set_filter_clauses_provider` is the old name of `set_filter_clauses_generator`).

- If the `FilterDataProvider` is registered with `add` like current, every time I get it from the container, I receive the new instance with a known default clauses generator. I gain the safety and readability but lose the cache. For filter blocks, because they use that default clauses generator, I can just get the instance and call the needed method:

https://github.com/woocommerce/woocommerce/blob/3f0cf66ba050a2258e79a03590bf7eaba9343ae5/plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterAttribute.php#L309-L309

- If the `FilterDataProvider` is registered with `shared`. I gain the cache but I have to set the generator every time I use `FilterDataProvider` to ensure I'm using the correct one.

```php
// StoreAPI/ProductQueryFitlers.php 
    $this->filter_data   = wc_get_container()->get( FilterDataProvider::class );
    $this->product_query = new ProductQuery();
    $this->filter_data->set_filter_clauses_generator( $this->product_query ); // Set the generator for StoreAPI
...
// BlockTypes/ProductFilterAttribute.php
    $provider = wc_get_container()->get( FilterDataProvider::class );
    // We need to set the clause generator again because it can be set to other before.
    $provider->set_filter_clauses_generator( wc_get_container()->get( FilterClausesGenerator::class ) );
    $count = $provider->get_attribute_counts( $query_vars, $slug );
```

Reply:
Ok, makes sense. But you could please a short comment explaining a summarized version of this close to the call to `add` in `register`? Otherwise someone missing this context might decide to change the `add` to a `share` during a refactor.

Reply:
Added in https://github.com/woocommerce/woocommerce/pull/43772/commits/116ee0475f4cdec1456ba0291bc0ef1d342ff821